File: plugin/src/core/reconcile.ts
export const STATUS_PREFIX = 'status/';
export const PRIORITY_PREFIX = 'priority/';

export interface Reconciliation {
  tags: Set<string>;
  autoTags: Set<string>;
  suppressedTags: Set<string>;
}

export interface ReconcileOptions {
  suppressedTags?: Set<string>;
  allowTagRemoval?: boolean;
}

/**
 * Reconcile current Zotero tags with accepted classification decisions.
 * Preserves human tags, suppresses deleted auto tags, and rejects status/priority mutations.
 */
export function reconcile(
  current: Set<string>,
  previousAuto: Set<string>,
  accepted: Set<string>,
  options: ReconcileOptions = {}
): Reconciliation {
  const { suppressedTags = new Set<string>(), allowTagRemoval = true } = options;

  // Workflow status & priority are exclusively human-owned
  const cleanPrevAuto = new Set(
    [...previousAuto].filter(
      (t) => !t.startsWith(STATUS_PREFIX) && !t.startsWith(PRIORITY_PREFIX)
    )
  );
  const cleanAccepted = new Set(
    [...accepted].filter(
      (t) => !t.startsWith(STATUS_PREFIX) && !t.startsWith(PRIORITY_PREFIX)
    )
  );
  const cleanSuppressed = new Set(
    [...suppressedTags].filter(
      (t) => !t.startsWith(STATUS_PREFIX) && !t.startsWith(PRIORITY_PREFIX)
    )
  );

  // Absence of a formerly-owned auto tag is a durable human suppression
  const deletedAuto = new Set([...cleanPrevAuto].filter((t) => !current.has(t)));
  const suppressed = new Set([...cleanSuppressed, ...deletedAuto]);

  const retainedAuto = new Set([...cleanPrevAuto].filter((t) => current.has(t)));
  // Tags that already exist but are not ours remain human-owned
  const newAuto = new Set([...cleanAccepted].filter((t) => !current.has(t)));

  const desiredAuto = new Set<string>();
  for (const tag of retainedAuto) {
    if (cleanAccepted.has(tag) && !suppressed.has(tag)) {
      desiredAuto.add(tag);
    }
  }
  for (const tag of newAuto) {
    if (!suppressed.has(tag)) {
      desiredAuto.add(tag);
    }
  }

  if (!allowTagRemoval) {
    for (const tag of retainedAuto) {
      desiredAuto.add(tag);
    }
  }

  // Final tags = (current - retainedAuto) | desiredAuto
  const humanTags = new Set([...current].filter((t) => !retainedAuto.has(t)));
  const finalTags = new Set([...humanTags, ...desiredAuto]);

  return {
    tags: finalTags,
    autoTags: desiredAuto,
    suppressedTags: suppressed,
  };
}

File: src/zotero_organiser/reconcile.py
from __future__ import annotations

from collections.abc import Set
from dataclasses import dataclass


STATUS_PREFIX = "status/"


@dataclass(frozen=True)
class Reconciliation:
    tags: set[str]
    auto_tags: set[str]
    suppressed_tags: set[str]


def reconcile(
    current: set[str],
    previous_auto: set[str],
    accepted: set[str],
    *,
    suppressed_tags: Set[str] = frozenset(),
    allow_tag_removal: bool = True,
) -> Reconciliation:
    # Workflow status is exclusively human-owned. Ignore any status ownership
    # left in state by an older release and defensively reject classifier output.
    previous_auto = {tag for tag in previous_auto if not tag.startswith(STATUS_PREFIX)}
    accepted = {tag for tag in accepted if not tag.startswith(STATUS_PREFIX)}
    suppressed_tags = {tag for tag in suppressed_tags if not tag.startswith(STATUS_PREFIX)}
    # Absence of a formerly-owned tag is a durable human suppression.
    suppressed = set(suppressed_tags) | (previous_auto - current)
    retained_auto = previous_auto & current
    # Tags that already exist but are not ours remain human-owned.  In
    # particular, seeing an accepted taxonomy tag must not make it ours.
    new_auto = accepted - current
    desired_auto = ((retained_auto & accepted) | new_auto) - suppressed
    if not allow_tag_removal:
        desired_auto |= retained_auto
    final = (current - retained_auto) | desired_auto
    return Reconciliation(final, desired_auto, suppressed)
